cmd/snap-update-ns/utils.go

```diff
@@ -0,0 +1,144 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+)
+
+// not available through syscall
+const (
+	UMOUNT_NOFOLLOW = 8
+)
+
+// For mocking everything during testing.
+var (
+	osLstat = os.Lstat
+
+	sysClose   = syscall.Close
+	sysMkdirat = syscall.Mkdirat
+	sysMount   = syscall.Mount
+	sysOpen    = syscall.Open
+	sysOpenat  = syscall.Openat
+	sysUnmount = syscall.Unmount
+	sysFchown  = syscall.Fchown
+
+	secureMkdirAll   = SecureMkdirAllImpl
+	ensureMountPoint = EnsureMountPointImpl
+)
+
+// SecureMkdirAll is the secure variant of os.MkdirAll.
+//
+// Unlike the regular version this implementation does not follow any symbolic
+// links. At all times the new directory segment is created using mkdirat(2)
+// while holding an open file descriptor to the parent directory.
+//
+// The only handled error is mkdirat(2) that fails with EEXIST. All other
+// errors are fatal but there is no attempt to undo anything that was created.
+//
+// The uid and gid are used for the fchown(2) system call which is performed
+// after each segment is created and opened. The special value -1 may be used
+// to request that ownership is not changed.
+func SecureMkdirAllImpl(name string, perm os.FileMode, uid, gid int) error {
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	var fd int
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+
+	if !path.IsAbs(name) {
```

Comment:
If you are going to make another commit, a comment here would be nice saying that we actively chose not to support relative directories. Eg:
```
// Only support absolute paths to avoid bugs in snap-confine when
// called from anywhere
```

Reply:
Sure, I'll add this before merging. Thank you!

Reply:
Done